Update publish.sh script documentation

kallimachos opened the pull request:
Clarify commands
Add setup checks
Add some troubleshooting

Closes #153

rackernexus commented:
:mag_right: [Preview](https://staging.developer.rackspace.com/build-82fbb09202/docs/glossary/glossary/)

klevenstein commented on doc/tools/internal-publish-script.rst:
```diff
@@ -84,6 +84,12 @@ or through VPN.
       upstream	git@github.com:rackerlabs/docs-rpc.git (fetch)
       upstream	git@github.com:rackerlabs/docs-rpc.git (push)
 
+#. Set the remote for master to upstream:
```
Might need to do `git fetch upstream` first

kallimachos replied:
As far as I know, the `git remote add` and `git branch --set-upstream-to` commands just alter the git config file. I think so long as we `git fetch upstream` prior to creating the local stable branches we should be ok. But if you hit an error at this point, we can always move the `git fetch upstream` command from the next step prior to this one.

klevenstein commented:
Other comments:

"Create local versions of the stable upstream branches:" - is there a need to check out `master` between creating each stable branch local version?

So I ran everything and ... 404'ed the site, apparently? Here's the output after I ran publish.sh: https://gist.github.com/klevenstein/4408b2e9de5a558935d9522456839b76

klevenstein commented:
I undid my screw-up by reverting the bad commit that wiped out gh-pages. All should be well for now, I hope?

meker12 commented:
Looking at the script output in the gist, it looks like the Sphinx build didn't run. So nothing was copied to the gh-pages branch.  The command to delete everything in the directory ran successfully, but the output from the Sphinx build didn't get copied to the directory.

klevenstein commented:
I suspected it might be something like that. I suspect Sphinx is borked on my machine since the re-image.

Time to re-install all the things...

kallimachos commented:
So yeah, Sphinx failing is a problem, because at the moment the script will carry on and push an empty repository, which leads to a 404 for the site. I had a similar issue yesterday when the pyenchant import started throwing errors that killed the Sphinx build. Since the spelling checks aren't critical, I just caught the exception and ignored it. See https://github.com/rackerlabs/docs-workstream/issues/86

We should trap Sphinx errors and exit the script if Sphinx fails. I've added this to https://github.com/rackerlabs/docs-workstream/issues/84. I've also suggested adding a check for the gh-pages site; if it comes back 404 then we roll back to the previous commit.

That said, I don't think trapping Sphinx errors would have helped in this case, as Sphinx appears not to have run at all. The critical line seems to be 826: `Agreeing to the Xcode/iOS license requires admin privileges, please re-run as root via sudo.` @klevenstein it looks like your suspicion is correct about the re-image. I'm guessing you need to allow Xcode on your machine in order for Sphinx to run.

Finally, the basename error comes from this line in the script:
```
[ "$(basename $0)" = "$(basename $BASH_SOURCE)" ] && tox=true || tox=false
```
It is there because you can't return from a bash script called directly, you can only exit. So we need to know if publish.sh is being sourced or run by tox. I don't know why it is throwing an error, but I'm getting the same thing as well. It looks like either $0 or $BASH_SOURCE is returning something unexpected. All that said, this error didn't cause the script to fail when I ran it, so to a certain extent we can ignore it for the moment. I've added it to my list of things to fix. :)

rackernexus commented:
:eyes: [Preview](https://staging.developer.rackspace.com/build-37757098d8/docs/glossary/glossary/)

rackernexus commented:
:eyes: [Preview](https://staging.developer.rackspace.com/build-6858c1d6fd/docs/glossary/glossary/)

kallimachos commented:
As per https://github.com/rackerlabs/docs-workstream/issues/84 I've added a check for Sphinx errors and removed the `basename` line.

With the updated instructions, there should be no need to checkout master in between creating the local stable branches. For example, `git checkout -b v12 upstream/v12` specifies that the new local branch should be created from upstream/v12.

Thanks for trying it out. Each time we hit a new error we get closer to making this work. :)

klevenstein commented:
LGTM. Thanks, and enjoy your time off, @kallimachos!
